### PR TITLE
#643: Implemented location information for expect statements

### LINKED ISSUE #643: expect statement message should include program location
This termination message from an expect statement

Program halted: expectation violation

should include a program location

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #643.

### CODE REVIEW

File: Test/expectations/Expect.dfy.expect [hunk 1]
```diff
@@ -1,12 +1,12 @@
 
 Dafny program verifier finished with 1 verified, 0 errors
-Program halted: expectation violation
+Program halted: Expect.dfy(21,2): expectation violation
```
Comment (robin-aws) [inline Test/expectations/Expect.dfy.expect:3]: super-small nit: I'm not a fan of the double colon in the output. :) I can't think of any precedent for this output from other language runtimes off the top of my head though, so I'm not sure exactly what to suggest.

**Comment by davidcok** [comment]
I concur. The problem is that the halt statement uses basic exceptions or built-in ‘panic’ calls to halt the program. The location string has to go in the accompanying message. To do otherwise might be possible for some target languages by writing a custom exception class that carried the location info. It would be helpful to merge this PR and if you want, keep the issue open to adjust the formatting.

**Comment by robin-aws** [comment]
I'm fine with the way the location is attached to the message. How about we change the "Program halted: " part instead?

```
[Program Halted] Expect.dfy(21,2): expectation violation
```

**Comment by davidcok** [comment]
Implemented and the PR ready to merge

**Comment by robin-aws** [comment]
Thanks for fixing the output format! Definitely a little cleaner.

One more test case we should have that I just thought of: the lit framework always passes `-useBaseNameForFileName` to ensure verification/compilation errors don't have full paths in them. With this implementation, if that option isn't passed does the token produce a full source path? This will be a little annoying to express as a lit test, but I had to do some similar post-processing for the TestAttribute.dfy test case.

**Comment by davidcok** [comment]
The filename in the message is the same as the filename in all messages - and is just the filename as stored in the token

**Comment by robin-aws** [comment]
I verified myself that the option is handled by the parser and therefore the token will have the correct string stored. Without the option you do indeed get the full(er) path:

```
a483e79acc1e:dafny salkeldr$ dafny /compile:3 Test/expectations/Expect.dfy
Dafny 2.3.0.10506

Dafny program verifier finished with 1 verified, 0 errors
Running...

[Program halted] Test/expectations/Expect.dfy(21,2): expectation violation
```

LGTM!

File: Test/expectations/Expect.dfy.expect (hunk 1 above)
Comment (davidcok) [reply Test/expectations/Expect.dfy.expect]: Format adjusted per a subsequent email
[Program halted] location: message

**Pull request merged**